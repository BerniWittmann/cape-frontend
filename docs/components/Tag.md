# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


