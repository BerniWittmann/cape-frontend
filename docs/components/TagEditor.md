# TagEditor

A Component to edit a list of tags

## Props

<!-- @vuese:TagEditor:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tags|The Array of Tag Objects|`Array`|`true`|-|

<!-- @vuese:TagEditor:props:end -->


## Events

<!-- @vuese:TagEditor:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Fired when the tag array changes|The new tag array|

<!-- @vuese:TagEditor:events:end -->


