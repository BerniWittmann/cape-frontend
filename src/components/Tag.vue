<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
