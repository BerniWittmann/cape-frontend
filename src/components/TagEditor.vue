<template>
  <div>
    <tag v-for="tag in selectedTags" :tag="tag" :key="tag.id" closable @close="removeTag(tag)"></tag>
    <el-select
            v-if="tagInputVisible"
            v-model="newTag"
            size="small"
            ref="tagSelect"
            @keyup.enter.native="addTag"
            @change="addTag"
            @blur="hideTagInput"
            filterable
            class="tag-space"
    >
      <el-option
              v-for="tag in availableTags"
              :key="tag.id"
              :label="tag.name"
              :value="tag">
        <tag :tag="tag" size="mini"></tag>
      </el-option>
    </el-select>
    <el-button v-if="!tagInputVisible && availableTags.length > 0" class="tag-space" size="small"
               @click="showTagInput" icon="el-icon-plus">{{$t('process.edit.add_tag') }}
    </el-button>
  </div>
</template>

<script>
import Tag from '@/components/Tag.vue'

/*
 * @vuese
 * @group Components
 *
 * A Component to edit a list of tags
 */
export default {
  name: 'TagEditor',
  model: {
    prop: 'tags',
    event: 'change'
  },

  props: {
    // The Array of Tag Objects
    tags: {
      type: Array,
      required: true
    }
  },

  components: {
    Tag
  },

  data() {
    return {
      selectedTags: [],
      tagInputVisible: false,
      newTag: undefined
    }
  },

  watch: {
    tags: {
      deep: true,
      immediate: true,
      handler: function (newTags) {
        this.selectedTags = newTags
      }
    }
  },

  computed: {
    allTags() {
      return this.$store.state.tag.tags
    },

    usedTagIDs() {
      return this.selectedTags.map((t) => t.id)
    },

    availableTags() {
      return this.allTags.filter(t => !this.usedTagIDs.includes(t.id))
    }
  },

  methods: {
    // @vuese
    // remove a tag
    // @arg the tag to be removes
    removeTag(tag) {
      this.selectedTags = this.selectedTags.filter((t) => t.id !== tag.id)
      // Fired when the tag array changes
      // @arg The new tag array
      this.$emit('change', this.selectedTags)
    },

    // @vuese
    // adds a new tag
    addTag() {
      if (!this.selectedTags.includes(this.newTag)) {
        this.selectedTags.push(this.newTag)
        // Fired when the tag array changes
        // @arg The new tag array
        this.$emit('change', this.selectedTags)
        this.$refs.tagSelect.blur()
      }
    },

    showTagInput() {
      this.tagInputVisible = true
      this.$nextTick(() => {
        this.$refs.tagSelect.focus()
      })
    },

    hideTagInput() {
      setTimeout(() => {
        this.tagInputVisible = false
        this.newTag = undefined
      }, 100)
    }
  }
}

</script>

<style scoped lang="scss">
.tag-space {
  margin-left: 5px;
}
</style>
