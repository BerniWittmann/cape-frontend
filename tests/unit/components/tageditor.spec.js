import { shallowMount, config, TransitionStub } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import { Select, Option, Button } from 'element-ui'

import TagEditor from '@/components/TagEditor.vue'
import Tag from '@/components/Tag.vue'

describe('Components', () => {
  describe('TagEditor', () => {
    let propsData
    let cmp
    let state
    config.stubs.transition = TransitionStub

    beforeEach(() => {
      propsData = {
        tags: [{
          id: '142',
          name: 'My Tag',
          color: '#FF0FF0'
        }, {
          id: '666',
          name: 'Tag From Hell',
          color: '#FF0000'
        }]
      }
      state = {
        tag: {
          tags: [{
            id: '142',
            name: 'My Tag',
            color: '#FF0FF0'
          }, {
            id: '666',
            name: 'Tag From Hell',
            color: '#FF0000'
          }, {
            id: '999',
            name: 'Last Tag',
            color: '#F0F0F0'
          }, {
            id: '1',
            name: 'First Tag',
            color: '#F1F300'
          }]
        }
      }
      render()
    })

    function render() {
      cmp = shallowMount(TagEditor, {
        i18n,
        propsData,
        mocks: {
          $store: {
            state
          }
        },
        stubs: {
          'ElSelect': Select,
          'ElOption': Option
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders the tags', () => {
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(2)

      for (let i = 0; i < 2; i++) {
        const tag = tags.at(i)
        expect(tag.props('tag')).toEqual(propsData.tags[i])
        expect(tag.props('closable')).toEqual(true)
      }
    })

    it('if no tags sets it renders no tags', () => {
      cmp.setProps({ tags: [] })
      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(0)
    })

    it('can remove a tag', () => {
      const expected = [propsData.tags[0]]
      const tag = cmp.findAll(Tag).at(1)
      console.error = jest.fn()
      tag.vm.$emit('close')
      console.error.mockRestore()

      expect(cmp.emitted().change).toBeTruthy()
      expect(cmp.emitted().change[0]).toEqual([expected])

      const tags = cmp.findAll(Tag)
      expect(tags.length).toEqual(1)
      expect(tags.at(0).props('tag')).toEqual(expected[0])
    })

    describe('it can add a tag', () => {
      jest.useFakeTimers()

      it('it has a button to add tags', () => {
        expect(cmp.contains(Button)).toBeTruthy()
      })

      it('can show the option on new tag button click', (done) => {
        const btn = cmp.find(Button)
        btn.vm.$emit('click')

        expect.assertions(2)
        cmp.vm.$nextTick(() => {
          expect(cmp.vm.tagInputVisible).toBeTruthy()
          expect(cmp.html()).toMatchSnapshot()
          done()
        })
      })
      it('only shows available tags', (done) => {
        const btn = cmp.find(Button)
        btn.vm.$emit('click')

        cmp.vm.$nextTick(() => {
          const options = cmp.findAll('.el-select-dropdown__item')
          expect(options.length).toEqual(2)
          for (let i = 0; i < 2; i++) {
            const tag = options.at(i).find(Tag)
            expect(tag.exists()).toBeTruthy()
            expect(tag.props('tag')).toEqual(state.tag.tags[2 + i])
          }
          done()
        })
      })
      it('adds the new tag on click', (done) => {
        const btn = cmp.find(Button)
        btn.vm.$emit('click')

        cmp.vm.$nextTick(() => {
          const option = cmp.findAll('.el-select-dropdown__item').at(1)
          expect(option.exists()).toBeTruthy()
          option.trigger('click')

          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            done()
          })
        })
      })
      it('adds a tag', () => {
        cmp.vm.$refs.tagSelect = { blur: jest.fn() }
        const expected = [state.tag.tags[0], state.tag.tags[1], state.tag.tags[3]]
        cmp.vm.newTag = state.tag.tags[3]
        cmp.vm.addTag()
        expect(cmp.vm.selectedTags).toEqual(expected)
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp.findAll(Tag).at(2).props('tag')).toEqual(state.tag.tags[3])
        expect(cmp.emitted().change).toBeTruthy()
        expect(cmp.emitted().change[0][0]).toEqual(expected)
        expect(cmp.vm.$refs.tagSelect.blur).toHaveBeenCalled()
      })
      it('adds a tag just once', () => {
        cmp.vm.newTag = cmp.vm.tags[0]
        cmp.vm.addTag()
        expect(cmp.vm.tags.length).toEqual(2)
        expect(cmp.emitted().change).toBeFalsy()
      })
      it('closes the dropdown after adding A Tag', (done) => {
        const btn = cmp.find(Button)
        btn.vm.$emit('click')

        cmp.vm.$nextTick(() => {
          const select = cmp.find(Select)
          select.vm.$emit('blur')

          jest.runOnlyPendingTimers()

          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            done()
          })
        })
      })
    })
    it('hides the add tag button if no tags are available', () => {
      state.tag.tags = [state.tag.tags[0], state.tag.tags[1]]
      render()

      expect(cmp.contains(Button)).toBeFalsy()
    })
  })
})
