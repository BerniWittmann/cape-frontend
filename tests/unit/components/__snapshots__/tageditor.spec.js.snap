// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components TagEditor it can add a tag adds a tag 1`] = `
<div>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <!---->
  <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="tag-space">process.edit.add_tag
  </elbutton-stub>
</div>
`;

exports[`Components TagEditor it can add a tag adds the new tag on click 1`] = `
<div>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <div class="el-select tag-space el-select--small">
    <!---->
    <div class="el-input el-input--small el-input--suffix is-focus">
      <!----><input type="text" autocomplete="off" placeholder="请选择" class="el-input__inner">
      <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up is-reverse"></i><!----><!----></span>
      <!----></span>
      <!---->
    </div>
    <div class="el-select-dropdown el-popper" style="" name="el-zoom-in-top">
      <div class="el-scrollbar" style="">
        <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
          <ul class="el-scrollbar__view el-select-dropdown__list">
            <!---->
            <li class="el-select-dropdown__item selected" style="">
              <tag-stub tag="[object Object]" size="mini"></tag-stub>
            </li>
            <li class="el-select-dropdown__item selected">
              <tag-stub tag="[object Object]" size="mini"></tag-stub>
            </li>
          </ul>
        </div>
        <div class="el-scrollbar__bar is-horizontal">
          <div class="el-scrollbar__thumb" style="transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
        </div>
        <div class="el-scrollbar__bar is-vertical">
          <div class="el-scrollbar__thumb" style="transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
        </div>
      </div>
      <!---->
    </div>
  </div>
  <!---->
</div>
`;

exports[`Components TagEditor it can add a tag can show the option on new tag button click 1`] = `
<div>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <div class="el-select tag-space el-select--small">
    <!---->
    <div class="el-input el-input--small el-input--suffix">
      <!----><input type="text" readonly="readonly" autocomplete="off" placeholder="请选择" class="el-input__inner">
      <!----><span class="el-input__suffix"><span class="el-input__suffix-inner"><i class="el-select__caret el-input__icon el-icon-arrow-up"></i><!----><!----></span>
      <!----></span>
      <!---->
    </div>
    <div class="el-select-dropdown el-popper" style="display: none;" name="el-zoom-in-top">
      <div class="el-scrollbar" style="">
        <div class="el-select-dropdown__wrap el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
          <ul class="el-scrollbar__view el-select-dropdown__list">
            <!---->
            <li class="el-select-dropdown__item">
              <tag-stub tag="[object Object]" size="mini"></tag-stub>
            </li>
            <li class="el-select-dropdown__item">
              <tag-stub tag="[object Object]" size="mini"></tag-stub>
            </li>
          </ul>
        </div>
        <div class="el-scrollbar__bar is-horizontal">
          <div class="el-scrollbar__thumb" style="width: 0px; transform: translateX(0%); -webkit-transform: translateX(0%);"></div>
        </div>
        <div class="el-scrollbar__bar is-vertical">
          <div class="el-scrollbar__thumb" style="height: 0px; transform: translateY(0%); -webkit-transform: translateY(0%);"></div>
        </div>
      </div>
      <!---->
    </div>
  </div>
  <!---->
</div>
`;

exports[`Components TagEditor it can add a tag closes the dropdown after adding A Tag 1`] = `
<div>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <!---->
  <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="tag-space">process.edit.add_tag
  </elbutton-stub>
</div>
`;

exports[`Components TagEditor renders 1`] = `
<div>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <tag-stub tag="[object Object]" closable="true"></tag-stub>
  <!---->
  <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="tag-space">process.edit.add_tag
  </elbutton-stub>
</div>
`;
